Inline init methods in the JS constructors.

sjrd opened the pull request:
This a rebased/updated version of #2142.

At definition site, instead of having a constructor that only creates fields, and an `init___XYZ` method for the actual constructor, we inline the body of the `init___XYZ` method inside the JavaScript constructor, giving:
```js
class $c_Foo extends $c_O {
  constructor(args) {
    super();
    this.x$1 = 0;
    this.y$1 = 0;
    // body of the init method
  }
}
```
At call site, this allows to emit
```js
new $c_Foo(args);
```
instead of
```js
new $c_Foo().init___XYZ(args);
```
This produces shorter code, and is potentially friendlier to JavaScript JITs.

We can do this for a class when all of the following apply:

* The class is not instantiated from `scalajsenv.js`
* It does not have any instantiated subclasses
* It has exactly one (regular) constructor
* It does not have any exported constructor

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4146/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4146/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5518/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5518/</a>
Test PASSed.

sjrd commented:
/cc @tOverney

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4148/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4148/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5520/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5520/</a>
Test PASSed.

gzm0 commented:
Any particular reason this is targeted on master?

sjrd commented:
Oh yes: not having to deal with Rhino and its own peculiar encoding of stuff!
IMO that's the biggest deal about dropping Rhino (besides 2x faster CI): the fact that it's much easier to improve the emitter.

gzm0 commented on tools/scalajsenv.js:
```diff
@@ -922,7 +922,7 @@ $TypeData.prototype.getClassOf = function() {
 //!endif
   if (!this._classOf)
-    this._classOf = new $c_jl_Class().init___jl_ScalaJSClassData(this);
+    this._classOf = new $c_jl_Class(this);
```
This contradicts the commit message :)

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -133,8 +133,43 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
   }
 
+  /** Extracts the inlined init constructor, if there is one. */
+  def extractInlinedInit(tree: LinkedClass)(
```
Consider "inlineable" (here and in many places). "inlined" sounds like it has already happened. If you are concerned about it sounding like a choice, at least improve the doc.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -133,8 +133,43 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
+  }
+
+  /** Generates an inlined init. */
+  def genInlinedInit(className: String, method: MethodDef)(
```
I'm not sure the general purpose method `genInlinedInit` brings any value here... It's a `genMethod` essentially :S

In any case, the name or the doc needs to improve. It is absolutely not clear what parts of the overall inlined constructor are generated. (and "Generates an inlined init" as a doc comment on `genInlineInit` is frankly useless).

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -133,8 +133,43 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
   /** Generates the JS constructor for a class. */
-  def genConstructor(tree: LinkedClass)(
+  def genConstructor(tree: LinkedClass,
+      inlinedInit: Option[WithGlobals[js.Function]])(
```
`initToInline`? It is not inlined (yet).

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala:
```diff
@@ -345,11 +345,20 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
+        val inlinedInitCache = classCache.getInlinedInitCache()
+        inlinedInitCache.getOrElseUpdate(m.version,
+            classEmitter.genInlinedInit(className, m.tree)(inlinedInitCache))
+      }
```
Not sure this is the best place to put all this. Have you considered putting all this inside `classEmitter` and simply filtering out the inlined constructor in the for loop below (the entire `emitter` knows about inlined constructors anyways, so it doesn't couple any tighter).

In any case, I do not see a use for the `inlinedInitCache` it gets cached right away afterwards...

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala:
```diff
@@ -345,11 +345,20 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
+      }
       val ctor = classTreeCache.constructor.getOrElseUpdate(
-          classEmitter.genConstructor(linkedClass)(classCache))
+          classEmitter.genConstructor(linkedClass, inlinedInit)(classCache))
```
This is broken. If `inlinedInit` gets updated (i.e. the ctor body), but the class does not, this does not regenerate.

sjrd replied:
Oh gosh. The whole point of the separate `genInlinedInit` and `inlinedInitCache` was to make this work in all cases ... and it's still broken. Well spotted!

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -73,6 +78,52 @@ private[emitter] final class KnowledgeGuardian {
+    )
+
+    val scalaClassDefs = linkingUnit.classDefs.filter(_.kind.isClass)
+    val instantiatedClasses = scalaClassDefs.filter(_.hasInstances)
```
Consider inlining this below to make it explicit that it is unused later. Being able to give it a name does not add any value IMO.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -73,6 +78,52 @@ private[emitter] final class KnowledgeGuardian {
+      if enableInlinedInitFor(classDef)
+    } yield {
+      classDef.encodedName
+    }).toSet
```
Use `classDef.filter(enableInlinedInitFor).map(_.encodedName).toSet`?

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -73,6 +78,52 @@ private[emitter] final class KnowledgeGuardian {
+       * - The class is not blacklisted
+       * - It does not have any instantiated subclass
+       * - It has exactly one (regular) constructor
+       * - It does not have an exported constructor
```
I like this list. I'd consider not constructing the intermediate set `cantHaveInlinedInit` and just add two set lookups.

gzm0 commented:
> Oh yes: not having to deal with Rhino and its own peculiar encoding of stuff!

Ah, ah, indeed.

gzm0 commented:
Also, just for my understanding: The restriction on exported constructors is that if we create an exported constructor on a class with a single constructor, we essentially create more than one constructor (the real one, and the exported one that typechecks). In ES6, we cannot forward constructors, so both of them need to be real constructors, therefore inlining is hardly desirable (since it would copy the init methods). Sounds about right?

sjrd commented:
Updated.

> The restriction on exported constructors is that if we create an exported constructor on a class with a single constructor, we essentially create more than one constructor (the real one, and the exported one that typechecks).

Yes, that's the idea. I added a small line about that in the list of criteria necessary for `init` to be inlineable.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala:
```diff
@@ -502,25 +529,38 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
-  private final class MethodCache extends knowledgeGuardian.KnowledgeAccessor {
-    private[this] var _tree: WithGlobals[js.Tree] = null
+  private final class MethodCache[TreeType]
+      extends knowledgeGuardian.KnowledgeAccessor {
```
I can actually revert that generalization with a type parameter. Now all uses of `MethodCache` are back to using `js.Tree`.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4149/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4149/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5522/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5522/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4150/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4150/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5523/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5523/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4151/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4151/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5524/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5524/</a>
Test PASSed.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -870,9 +945,7 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
-            js.Apply(
-                js.New(encodeClassVar(className), Nil) DOT js.Ident("init___"),
-                Nil)
+            js.New(encodeClassVar(className), Nil)
```
This implicitly uses the fact that every module class only has a single init method, right? I'm fine with imposing this, but is it IR checked? (Before, one could theoretically add another constructor and instantiate the class by hand, not that it is a good idea).

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala:
```diff
@@ -444,6 +460,11 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
 
+  private def mergeVersions(v1: Option[String],
+      v2: Option[String]): Option[String] = {
+    v1.flatMap(s1 => v2.map(s2 => s1 + s2))
```
Nope :( Consider:

```scala
mergeVersions("1", "11") // --> 111
mergeVersions("11", "1") // --> 111
```

Bad. Maybe we should allow `Any` as version and just rely on the equality method? Then you could just use a tuple here? (or do we need versions to be serializable?)

gzm0 replied:
We probably could allow `Any` versions in the caches?

sjrd replied:
I think here we can solve this case with `s1.length + "-" + s1 + s2`.

gzm0 replied:
That works too, yes.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -73,6 +78,57 @@ private[emitter] final class KnowledgeGuardian {
+       * - It does not have any exported constructor (since they are
+       *   effectively secondary constructors)
+       *
+       * By construction, this is always true for module classes.
```
IR check this :)

sjrd commented:
Updated.

Note that, although I do check that module classes have exactly 1 constructor (without parameter), I do not check yet that module classes are not *extended*. Doing this proves to be quite difficult, and reveals that inheritance relationships between classes of all kinds are woefully underchecked. This needs a dedicated PR.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala:
```diff
@@ -345,11 +345,27 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
+         */
+        val ctorCache = classCache.getConstructorCache()
+        val optInlineableInitVersion =
+          linkedInlineableInit.fold[Option[String]](Some(""))(_.version)
```
Oh man, I also didn't spot this... In theory this can also cause a collision (if the method has version `""`).

sjrd replied:
Yeah but ... that's not really something we do, having `""` as version. We might even have some implicit guarantee that a version is never `Some("")`.

If you prefer, it's still possible to be more robust with something like:
```scala
ctorVersion = linkedInlineableInit.fold[Option[String]] {
  linkedClass.version.map("1-" + _)
} { inlineableInitVersion =>
  mergeVersions(linkedClass.version, inlineableInitVersion).map("2-" + _)
}
```

gzm0 replied:
I'd rather have that, yes. Sorry for not spotting it earlier. The problem with implicit guarantees is that they are implicit :)

sjrd replied:
Done.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4152/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4152/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5526/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5526/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4153/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4153/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5527/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5527/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4154/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4154/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5528/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5528/</a>
Test PASSed.
